File: app/src/main/java/com/krystianwsul/checkme/gui/instances/tree/DividerNode.java
package com.krystianwsul.checkme.gui.instances.tree;

import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.view.View;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.utils.InstanceKey;
import com.krystianwsul.treeadapter.ModelNode;
import com.krystianwsul.treeadapter.NodeContainer;
import com.krystianwsul.treeadapter.TreeNode;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

class DividerNode extends GroupHolderNode implements ModelNode {
    @NonNull
    private final NodeCollection mNodeCollection;

    private TreeNode mTreeNode;

    @NonNull
    private final ArrayList<DoneInstanceNode> mDoneInstanceNodes = new ArrayList<>();

    DividerNode(float density, int indentation, @NonNull NodeCollection nodeCollection) {
        super(density, indentation);

        mNodeCollection = nodeCollection;
    }

    @NonNull
    TreeNode initialize(boolean expanded, @NonNull NodeContainer nodeContainer, @NonNull List<GroupListFragment.InstanceData> doneInstanceDatas, @Nullable HashMap<InstanceKey, Boolean> expandedInstances) {
        Assert.assertTrue(!expanded || !doneInstanceDatas.isEmpty());

        mTreeNode = new TreeNode(this, nodeContainer, expanded, false);

        List<TreeNode> childTreeNodes = Stream.of(doneInstanceDatas)
                .map(doneInstanceData -> newChildTreeNode(doneInstanceData, expandedInstances))
                .collect(Collectors.toList());

        mTreeNode.setChildTreeNodes(childTreeNodes);

        return mTreeNode;
    }

    @NonNull
    private TreeNode newChildTreeNode(@NonNull GroupListFragment.InstanceData instanceData, @Nullable HashMap<InstanceKey, Boolean> expandedInstances) {
        Assert.assertTrue(instanceData.Done != null);

        DoneInstanceNode doneInstanceNode = new DoneInstanceNode(mDensity, mIndentation, instanceData, this);

        TreeNode childTreeNode = doneInstanceNode.initialize(mTreeNode, expandedInstances);

        mDoneInstanceNodes.add(doneInstanceNode);

        return childTreeNode;
    }

    boolean expanded() {
        TreeNode treeNode = getTreeNode();

        return treeNode.expanded();
    }

    void addExpandedInstances(@NonNull HashMap<InstanceKey, Boolean> expandedInstances) {
        for (DoneInstanceNode doneInstanceNode : mDoneInstanceNodes)
            doneInstanceNode.addExpandedInstances(expandedInstances);
    }

    @Override
    int getNameVisibility() {
        return View.VISIBLE;
    }

    @NonNull
    @Override
    String getName() {
        return getGroupListFragment().getString(R.string.done);
    }

    @Override
    int getNameColor() {
        return ContextCompat.getColor(getGroupListFragment().getActivity(), R.color.textPrimary);
    }

    @Override
    boolean getNameSingleLine() {
        return true;
    }

    @Override
    int getDetailsVisibility() {
        return View.GONE;
    }

    @NonNull
    @Override
    String getDetails() {
        throw new UnsupportedOperationException();
    }

    @Override
    int getDetailsColor() {
        throw new UnsupportedOperationException();
    }

    @Override
    int getChildrenVisibility() {
        return View.GONE;
    }

    @NonNull
    @Override
    String getChildren() {
        throw new UnsupportedOperationException();
    }

    @Override
    int getChildrenColor() {
        throw new UnsupportedOperationException();
    }

    @Override
    int getExpandVisibility() {
        return View.VISIBLE;
    }

    @Override
    int getExpandImageResource() {
        Assert.assertTrue(mTreeNode != null);

        if (mTreeNode.expanded())
            return R.drawable.ic_expand_less_black_36dp;
        else
            return R.drawable.ic_expand_more_black_36dp;
    }

    @NonNull
    @Override
    View.OnClickListener getExpandOnClickListener() {
        Assert.assertTrue(mTreeNode != null);

        return mTreeNode.getExpandListener();
    }

    @Override
    int getCheckBoxVisibility() {
        return View.INVISIBLE;
    }

    @Override
    boolean getCheckBoxChecked() {
        throw new UnsupportedOperationException();
    }

    @NonNull
    @Override
    View.OnClickListener getCheckBoxOnClickListener() {
        throw new UnsupportedOperationException();
    }

    @Override
    int getSeparatorVisibility() {
        TreeNode treeNode = getTreeNode();
        return (treeNode.getSeparatorVisibility() ? View.VISIBLE : View.INVISIBLE);
    }

    @Override
    int getBackgroundColor() {
        return Color.TRANSPARENT;
    }

    @NonNull
    @Override
    View.OnLongClickListener getOnLongClickListener() {
        TreeNode treeNode = getTreeNode();

        return treeNode.getOnLongClickListener();
    }

    @NonNull
    @Override
    View.OnClickListener getOnClickListener() {
        TreeNode treeNode = getTreeNode();

        return treeNode.getOnClickListener();
    }

    public void remove(@NonNull DoneInstanceNode doneInstanceNode) {
        Assert.assertTrue(mTreeNode != null);

        Assert.assertTrue(mDoneInstanceNodes.contains(doneInstanceNode));
        mDoneInstanceNodes.remove(doneInstanceNode);

        mTreeNode.remove(doneInstanceNode.getTreeNode());
    }

    public void add(@NonNull GroupListFragment.InstanceData instanceData) {
        Assert.assertTrue(mTreeNode != null);

        mTreeNode.add(newChildTreeNode(instanceData, null));
    }

    @NonNull
    NodeCollection getNodeCollection() {
        return mNodeCollection;
    }

    @NonNull
    GroupListFragment.GroupAdapter getGroupAdapter() {
        return getNodeCollection().getGroupAdapter();
    }

    @NonNull
    private GroupListFragment getGroupListFragment() {
        return getGroupAdapter().mGroupListFragment;
    }

    @Override
    public boolean selectable() {
        return false;
    }

    @Override
    public void onClick() {

    }

    @Override
    public int compareTo(@NonNull ModelNode another) {
        Assert.assertTrue(another instanceof NoteNode || another instanceof NotDoneGroupNode || another instanceof UnscheduledNode);
        return 1;
    }

    @Override
    public boolean visibleWhenEmpty() {
        return false;
    }

    @Override
    public boolean visibleDuringActionMode() {
        return false;
    }

    @Override
    public boolean separatorVisibleWhenNotExpanded() {
        return false;
    }

    @NonNull
    private TreeNode getTreeNode() {
        Assert.assertTrue(mTreeNode != null);

        return mTreeNode;
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/instances/tree/DividerNode.kt
package com.krystianwsul.checkme.gui.instances.tree

import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.domainmodel.GroupTypeFactory
import com.krystianwsul.checkme.gui.tree.*
import com.krystianwsul.checkme.gui.tree.delegates.expandable.ExpandableDelegate
import com.krystianwsul.checkme.gui.tree.delegates.indentation.IndentationDelegate
import com.krystianwsul.checkme.gui.tree.delegates.indentation.IndentationModelNode
import com.krystianwsul.checkme.gui.tree.delegates.singleline.SingleLineDelegate
import com.krystianwsul.checkme.gui.tree.delegates.singleline.SingleLineModelNode
import com.krystianwsul.checkme.gui.utils.flatten
import com.krystianwsul.treeadapter.ModelNode
import com.krystianwsul.treeadapter.NodeContainer
import com.krystianwsul.treeadapter.TreeNode

class DividerNode(
        override val indentation: Int,
        val nodeCollection: NodeCollection,
        override val parentNode: ModelNode<AbstractHolder>?,
) : AbstractModelNode(), SingleLineModelNode, IndentationModelNode {

    override val holderType = HolderType.EXPANDABLE_SINGLELINE

    override val id get() = Id(nodeCollection.nodeContainer.id)

    data class Id(val id: Any)

    override lateinit var treeNode: TreeNode<AbstractHolder>
        private set

    private val doneInstanceNodes = ArrayList<DoneInstanceNode>()

    private val groupAdapter get() = nodeCollection.groupAdapter

    private val groupListFragment get() = groupAdapter.groupListFragment

    override val delegates by lazy {
        listOf(
            ExpandableDelegate(treeNode),
            SingleLineDelegate(this),
            IndentationDelegate(this),
        )
    }

    override val disableRipple = true

    fun initialize(
        initialExpansionState: TreeNode.ExpansionState?,
        nodeContainer: NodeContainer<AbstractHolder>,
        doneSingleBridges: List<GroupTypeFactory.SingleBridge>,
        contentDelegateStates: Map<NotDoneNode.ContentDelegate.Id, NotDoneNode.ContentDelegate.State>,
    ): TreeNode<AbstractHolder> {
        treeNode = TreeNode(
            this,
            nodeContainer,
            initialExpansionState = initialExpansionState.takeIf { doneSingleBridges.isNotEmpty() },
        )

        val childTreeNodes = doneSingleBridges.map { newChildTreeNode(it, contentDelegateStates) }

        treeNode.setChildTreeNodes(childTreeNodes)

        return treeNode
    }

    private fun newChildTreeNode(
        singleBridge: GroupTypeFactory.SingleBridge,
        contentDelegateStates: Map<NotDoneNode.ContentDelegate.Id, NotDoneNode.ContentDelegate.State>,
    ): TreeNode<AbstractHolder> {
        checkNotNull(singleBridge.instanceData.done)

        val doneInstanceNode = DoneInstanceNode(indentation, singleBridge, this)

        val childTreeNode = doneInstanceNode.initialize(contentDelegateStates, treeNode)

        doneInstanceNodes.add(doneInstanceNode)

        return childTreeNode
    }

    val expansionState get() = treeNode.getSaveExpansionState()

    val contentDelegateStates get() = doneInstanceNodes.map { it.contentDelegate.states }.flatten()

    override val text by lazy { groupListFragment.activity.getString(R.string.done) }

    override fun compareTo(other: ModelNode<AbstractHolder>): Int {
        check(
            other is NotDoneGroupNode
                    || other is UnscheduledNode
                    || other is ImageNode
                    || other is DetailsNode
        )

        if (nodeCollection.doneBeforeNotDone && other is NotDoneGroupNode) return -1

        return 1
    }

    override fun isVisible(actionMode: Boolean, hasVisibleChildren: Boolean): Boolean {
        return hasVisibleChildren
    }
}
